Stream errors when response is sent early

In Akka 2.4.2-RC1, getting the following error when a response is sent before the request body has been read:

```
[WARN] [01/27/2016 12:19:19.448] [application-akka.actor.default-dispatcher-11] [akka.actor.ActorSystemImpl(application)] Sending 2xx response before end of request was received...
Note that the connection will be closed after this response. Also, many clients will not read early responses!
Consider waiting for the request end before dispatching this response!
[ERROR] [01/27/2016 12:19:19.475] [application-akka.actor.default-dispatcher-6] [akka.actor.ActorSystemImpl(application)] Outgoing response stream error
java.lang.IllegalArgumentException: requirement failed: Cannot push port (SubSource.out) twice
	at scala.Predef$.require(Predef.scala:219)
	at akka.stream.stage.GraphStageLogic.push(GraphStage.scala:439)
	at akka.stream.impl.fusing.SubSource$$anon$3$$anonfun$2.apply(StreamOfStreams.scala:519)
	at akka.stream.impl.fusing.SubSource$$anon$3$$anonfun$2.apply(StreamOfStreams.scala:516)
	at akka.stream.impl.fusing.GraphInterpreter.runAsyncInput(GraphInterpreter.scala:557)
	at akka.stream.impl.fusing.GraphInterpreterShell.receive(ActorGraphInterpreter.scala:399)
	at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:549)
	at akka.actor.Actor$class.aroundReceive(Actor.scala:480)
	at akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:495)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:526)
	at akka.actor.ActorCell.invoke(ActorCell.scala:495)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
	at akka.dispatch.Mailbox.run(Mailbox.scala:224)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

Also getting this, somewhat similar but different error, triggered in a different test:

```
[WARN] [01/27/2016 12:19:20.544] [application-akka.actor.default-dispatcher-7] [akka.actor.ActorSystemImpl(application)] Sending 2xx response before end of request was received...
Note that the connection will be closed after this response. Also, many clients will not read early responses!
Consider waiting for the request end before dispatching this response!
[ERROR] [01/27/2016 12:19:20.563] [application-akka.actor.default-dispatcher-8] [akka.actor.ActorSystemImpl(application)] Internal server error, sending 500 response
java.lang.IllegalArgumentException: requirement failed: Cannot pull port (RequestStartThenRunIgnore.in) twice
	at scala.Predef$.require(Predef.scala:219)
	at akka.stream.stage.GraphStageLogic.pull(GraphStage.scala:351)
	at akka.http.impl.engine.server.HttpServerBluePrint$PrepareRequests$$anon$4$$anon$6.onPull(HttpServerBluePrint.scala:123)
	at akka.stream.stage.GraphStageLogic$SubSourceOutlet$$anonfun$5.apply(GraphStage.scala:1027)
	at akka.stream.stage.GraphStageLogic$SubSourceOutlet$$anonfun$5.apply(GraphStage.scala:1023)
	at akka.stream.impl.fusing.GraphInterpreter.runAsyncInput(GraphInterpreter.scala:557)
	at akka.stream.impl.fusing.GraphInterpreterShell.receive(ActorGraphInterpreter.scala:399)
	at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:549)
	at akka.actor.Actor$class.aroundReceive(Actor.scala:480)
	at akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:495)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:526)
	at akka.actor.ActorCell.invoke(ActorCell.scala:495)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
	at akka.dispatch.Mailbox.run(Mailbox.scala:224)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
	at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

--- a/akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala
+++ b/akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala
@@ -100,38 +100,42 @@ private[http] object HttpServerBluePrint {
     val out = Outlet[HttpRequest]("RequestStartThenRunIgnore.out")
     override val shape: FlowShape[RequestOutput, HttpRequest] = FlowShape.of(in, out)
 
-    override def createLogic(inheritedAttributes: Attributes) = new GraphStageLogic(shape) {
+    override def createLogic(inheritedAttributes: Attributes) = new GraphStageLogic(shape) with InHandler with OutHandler {
       val remoteAddress = inheritedAttributes.get[HttpAttributes.RemoteAddress].flatMap(_.address)
-      var upstreamPullWaiting = false
+      var downstreamPullWaiting = false
+      var completionDeferred = false
 
-      val idle = new InHandler {
-        def onPush(): Unit = grab(in) match {
-          case RequestStart(method, uri, protocol, hdrs, entityCreator, _, _) ⇒
-            val effectiveMethod = if (method == HttpMethods.HEAD && settings.transparentHeadRequests) HttpMethods.GET else method
-            val effectiveHeaders =
-              if (settings.remoteAddressHeader && remoteAddress.isDefined)
-                headers.`Remote-Address`(RemoteAddress(remoteAddress.get)) +: hdrs
-              else hdrs
+      // optimization: to avoid allocations the "idle" case in and out handlers are put directly on the GraphStageLogic itself
+      override def onPull(): Unit = {
+        pull(in)
+      }
 
-            val entity = createEntity(entityCreator) withSizeLimit settings.parserSettings.maxContentLength
-            push(out, HttpRequest(effectiveMethod, uri, effectiveHeaders, entity, protocol))
-          case other ⇒
-            throw new IllegalStateException(s"unexpected element of type ${other.getClass}")
-        }
+      override def onPush(): Unit = grab(in) match {
+        case RequestStart(method, uri, protocol, hdrs, entityCreator, _, _) ⇒
+          val effectiveMethod = if (method == HttpMethods.HEAD && settings.transparentHeadRequests) HttpMethods.GET else method
+          val effectiveHeaders =
+            if (settings.remoteAddressHeader && remoteAddress.isDefined)
+              headers.`Remote-Address`(RemoteAddress(remoteAddress.get)) +: hdrs
+            else hdrs
+
+          val entity = createEntity(entityCreator) withSizeLimit settings.parserSettings.maxContentLength
+          push(out, HttpRequest(effectiveMethod, uri, effectiveHeaders, entity, protocol))
+        case other ⇒
+          throw new IllegalStateException(s"unexpected element of type ${other.getClass}")
       }
 
       setIdleHandlers()
 
       def setIdleHandlers() {
-        setHandler(in, idle)
-        setHandler(out, new OutHandler {
-          override def onPull(): Unit = {
+        if (completionDeferred) {
+          completeStage()
+        } else {
+          setHandler(in, this)
+          setHandler(out, this)
+          if (downstreamPullWaiting) {
+            downstreamPullWaiting = false
             pull(in)
           }
-        })
-        if (upstreamPullWaiting) {
-          upstreamPullWaiting = false
-          pull(in)
         }
       }
 
@@ -143,13 +147,13 @@ private[http] object HttpServerBluePrint {
         }
 
       def streamRequestEntity(creator: (Source[ParserOutput.RequestOutput, NotUsed]) => RequestEntity): RequestEntity = {
-        // stream the request entity until we reach the end of it
+        // stream incoming chunks into the request entity until we reach the end of it
+        // and then toggle back to "idle"
+
         val entitySource = new SubSourceOutlet[RequestOutput]("EntitySource")
-        entitySource.setHandler(new OutHandler {
-          def onPull(): Unit = {
-            pull(in)
-          }
-        })
+        // optimization: re-use the idle outHandler
+        entitySource.setHandler(this)
+
         setHandler(in, new InHandler {
           def onPush(): Unit = {
             grab(in) match {
@@ -160,13 +164,26 @@ private[http] object HttpServerBluePrint {
               case x ⇒ entitySource.push(x)
             }
           }
-          override def onUpstreamFinish(): Unit = completeStage()
+          override def onUpstreamFinish(): Unit = {
+            entitySource.complete()
+            completeStage()
+          }
+          override def onUpstreamFailure(ex: Throwable): Unit = {
+            entitySource.fail(ex)
+            failStage(ex)
+          }
         })
         setHandler(out, new OutHandler {
           override def onPull(): Unit = {
-            // remember this until we are done with the entity
-            // so we can pass it downstream at that point
-            upstreamPullWaiting = true
+            // remember this until we are done with the chunked entity
+            // so can pull downstream then
+            downstreamPullWaiting = true
+          }
+          override def onDownstreamFinish(): Unit = {
+            // downstream signalled not wanting any more requests
+            // we should keep processing the entity stream and then
+            // when it completes complete the stage
+            completionDeferred = true
           }
         })
         creator(Source.fromGraph(entitySource.source))

--- a/akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala
+++ b/akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala
@@ -89,6 +89,12 @@ private[http] object HttpServerBluePrint {
       case _                 ⇒ BidiFlow.identity
     }
 
+  /**
+   * Two state stage, either transforms an incoming RequestOutput into a HttpRequest with strict entity and then pushes
+   * that (the "idle" inHandler) or creates a HttpRequest with a streamed entity and switch to a state which will push
+   * incoming chunks into the streaming entity until end of request is reached (the StreamedEntityCreator case in create
+   * entity).
+   */
   final class PrepareRequests(settings: ServerSettings) extends GraphStage[FlowShape[RequestOutput, HttpRequest]] {
     val in = Inlet[RequestOutput]("RequestStartThenRunIgnore.in")
     val out = Outlet[HttpRequest]("RequestStartThenRunIgnore.out")
@@ -96,6 +102,7 @@ private[http] object HttpServerBluePrint {
 
     override def createLogic(inheritedAttributes: Attributes) = new GraphStageLogic(shape) {
       val remoteAddress = inheritedAttributes.get[HttpAttributes.RemoteAddress].flatMap(_.address)
+      var upstreamPullWaiting = false
 
       val idle = new InHandler {
         def onPush(): Unit = grab(in) match {
@@ -112,31 +119,59 @@ private[http] object HttpServerBluePrint {
             throw new IllegalStateException(s"unexpected element of type ${other.getClass}")
         }
       }
-      setHandler(in, idle)
+
+      setIdleHandlers()
+
+      def setIdleHandlers() {
+        setHandler(in, idle)
+        setHandler(out, new OutHandler {
+          override def onPull(): Unit = {
+            pull(in)
+          }
+        })
+        if (upstreamPullWaiting) {
+          upstreamPullWaiting = false
+          pull(in)
+        }
+      }
+
 
       def createEntity(creator: EntityCreator[RequestOutput, RequestEntity]): RequestEntity =
         creator match {
           case StrictEntityCreator(entity) ⇒ entity
-          case StreamedEntityCreator(creator) ⇒
-            val entitySource = new SubSourceOutlet[RequestOutput]("EntitySource")
-            entitySource.setHandler(new OutHandler {
-              def onPull(): Unit = pull(in)
-            })
-            setHandler(in, new InHandler {
-              def onPush(): Unit = grab(in) match {
-                case MessageEnd ⇒
-                  entitySource.complete()
-                  setHandler(in, idle)
-                case x ⇒ entitySource.push(x)
-              }
-              override def onUpstreamFinish(): Unit = completeStage()
-            })
-            creator(Source.fromGraph(entitySource.source))
+          case StreamedEntityCreator(creator) ⇒ streamRequestEntity(creator)
         }
 
-      setHandler(out, new OutHandler {
-        override def onPull(): Unit = pull(in)
-      })
+      def streamRequestEntity(creator: (Source[ParserOutput.RequestOutput, NotUsed]) => RequestEntity): RequestEntity = {
+        // stream the request entity until we reach the end of it
+        val entitySource = new SubSourceOutlet[RequestOutput]("EntitySource")
+        entitySource.setHandler(new OutHandler {
+          def onPull(): Unit = {
+            pull(in)
+          }
+        })
+        setHandler(in, new InHandler {
+          def onPush(): Unit = {
+            grab(in) match {
+              case MessageEnd ⇒
+                entitySource.complete()
+                setIdleHandlers()
+
+              case x ⇒ entitySource.push(x)
+            }
+          }
+          override def onUpstreamFinish(): Unit = completeStage()
+        })
+        setHandler(out, new OutHandler {
+          override def onPull(): Unit = {
+            // remember this until we are done with the entity
+            // so we can pass it downstream at that point
+            upstreamPullWaiting = true
+          }
+        })
+        creator(Source.fromGraph(entitySource.source))
+      }
+
     }
   }
 

--- a/akka-http-core/src/test/scala/akka/http/impl/engine/server/PrepareRequestsSpec.scala
+++ b/akka-http-core/src/test/scala/akka/http/impl/engine/server/PrepareRequestsSpec.scala
@@ -1,0 +1,104 @@
+/*
+ * Copyright (C) 2016 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.http.impl.engine.server
+
+import akka.http.impl.engine.parsing.ParserOutput
+import akka.http.impl.engine.parsing.ParserOutput.{ StrictEntityCreator, EntityStreamError, EntityChunk, StreamedEntityCreator }
+import akka.http.impl.engine.server.HttpServerBluePrint.PrepareRequests
+import akka.http.scaladsl.model._
+import akka.http.scaladsl.settings.ServerSettings
+import akka.stream.{ Attributes, ActorMaterializer }
+import akka.stream.scaladsl.{ Sink, Source, Flow }
+import akka.stream.testkit.{ TestSubscriber, TestPublisher }
+import akka.testkit.AkkaSpec
+import akka.util.ByteString
+import org.scalatest.{ Matchers, WordSpec }
+import scala.concurrent.duration._
+
+class PrepareRequestsSpec extends AkkaSpec {
+
+  "The PrepareRequest stage" should {
+
+    "not fail when there is demand from both streamed entity consumption and regular flow" in {
+      implicit val materializer = ActorMaterializer()
+      // covers bug #19623 where a reply before the streamed
+      // body has been consumed causes pull/push twice
+      val inProbe = TestPublisher.manualProbe[ParserOutput.RequestOutput]()
+      val upstreamProbe = TestSubscriber.manualProbe[HttpRequest]()
+
+      val stage = Flow.fromGraph(new PrepareRequests(ServerSettings(system)))
+
+      Source.fromPublisher(inProbe)
+        .via(stage)
+        .to(Sink.fromSubscriber(upstreamProbe))
+        .withAttributes(Attributes.inputBuffer(1, 1))
+        .run()
+
+      val upstreamSub = upstreamProbe.expectSubscription()
+      val inSub = inProbe.expectSubscription()
+
+      // let request with streamed entity through
+      upstreamSub.request(1)
+      inSub.expectRequest(1)
+      inSub.sendNext(ParserOutput.RequestStart(
+        HttpMethods.GET,
+        Uri("http://example.com/"),
+        HttpProtocols.`HTTP/1.1`,
+        List(),
+        StreamedEntityCreator[ParserOutput, RequestEntity] { entityChunks ⇒
+          val chunks = entityChunks.collect {
+            case EntityChunk(chunk)      ⇒ chunk
+            case EntityStreamError(info) ⇒ throw EntityStreamException(info)
+          }
+          HttpEntity.Chunked(ContentTypes.`application/octet-stream`, HttpEntity.limitableChunkSource(chunks))
+        },
+        true,
+        false))
+
+      val request = upstreamProbe.expectNext()
+
+      // and subscribe to it's streamed entity
+      val entityProbe = TestSubscriber.manualProbe[ByteString]()
+      request.entity.dataBytes.to(Sink.fromSubscriber(entityProbe))
+        .withAttributes(Attributes.inputBuffer(1, 1))
+        .run()
+
+      val entitySub = entityProbe.expectSubscription()
+
+      // the bug happens when both the client has signalled demand
+      // and the the streamed entity has
+      upstreamSub.request(1)
+      entitySub.request(1)
+
+      // then comes the next chunk from the actual request
+      inSub.expectRequest(1)
+
+      // bug would fail stream here with exception
+      upstreamProbe.expectNoMsg(100.millis)
+
+      inSub.sendNext(ParserOutput.EntityChunk(HttpEntity.ChunkStreamPart(ByteString("abc"))))
+      entityProbe.expectNext()
+      entitySub.request(1)
+      inSub.sendNext(ParserOutput.MessageEnd)
+      entityProbe.expectComplete()
+
+      // the rest of the test covers the saved pull
+      // that should go downstream when the streamed entity
+      // has reached it's end
+      inSub.expectRequest(1)
+      inSub.sendNext(ParserOutput.RequestStart(
+        HttpMethods.GET,
+        Uri("http://example.com/"),
+        HttpProtocols.`HTTP/1.1`,
+        List(),
+        StrictEntityCreator(HttpEntity.Strict(ContentTypes.`application/octet-stream`, ByteString("body"))),
+        true,
+        false))
+
+      upstreamProbe.expectNext()
+
+    }
+  }
+
+}

--- a/akka-http-core/src/test/scala/akka/http/impl/engine/server/PrepareRequestsSpec.scala
+++ b/akka-http-core/src/test/scala/akka/http/impl/engine/server/PrepareRequestsSpec.scala
@@ -18,6 +18,38 @@ import scala.concurrent.duration._
 
 class PrepareRequestsSpec extends AkkaSpec {
 
+  val chunkedStart =
+    ParserOutput.RequestStart(
+      HttpMethods.GET,
+      Uri("http://example.com/"),
+      HttpProtocols.`HTTP/1.1`,
+      List(),
+      StreamedEntityCreator[ParserOutput, RequestEntity] { entityChunks ⇒
+        val chunks = entityChunks.collect {
+          case EntityChunk(chunk)      ⇒ chunk
+          case EntityStreamError(info) ⇒ throw EntityStreamException(info)
+        }
+        HttpEntity.Chunked(ContentTypes.`application/octet-stream`, HttpEntity.limitableChunkSource(chunks))
+      },
+      true,
+      false)
+
+  val chunkPart =
+    ParserOutput.EntityChunk(HttpEntity.ChunkStreamPart(ByteString("abc")))
+
+  val chunkRequestComplete =
+    ParserOutput.MessageEnd
+
+  val strictRequest =
+    ParserOutput.RequestStart(
+      HttpMethods.GET,
+      Uri("http://example.com/"),
+      HttpProtocols.`HTTP/1.1`,
+      List(),
+      StrictEntityCreator(HttpEntity.Strict(ContentTypes.`application/octet-stream`, ByteString("body"))),
+      true,
+      false)
+
   "The PrepareRequest stage" should {
 
     "not fail when there is demand from both streamed entity consumption and regular flow" in {
@@ -41,20 +73,7 @@ class PrepareRequestsSpec extends AkkaSpec {
       // let request with streamed entity through
       upstreamSub.request(1)
       inSub.expectRequest(1)
-      inSub.sendNext(ParserOutput.RequestStart(
-        HttpMethods.GET,
-        Uri("http://example.com/"),
-        HttpProtocols.`HTTP/1.1`,
-        List(),
-        StreamedEntityCreator[ParserOutput, RequestEntity] { entityChunks ⇒
-          val chunks = entityChunks.collect {
-            case EntityChunk(chunk)      ⇒ chunk
-            case EntityStreamError(info) ⇒ throw EntityStreamException(info)
-          }
-          HttpEntity.Chunked(ContentTypes.`application/octet-stream`, HttpEntity.limitableChunkSource(chunks))
-        },
-        true,
-        false))
+      inSub.sendNext(chunkedStart)
 
       val request = upstreamProbe.expectNext()
 
@@ -87,16 +106,101 @@ class PrepareRequestsSpec extends AkkaSpec {
       // that should go downstream when the streamed entity
       // has reached it's end
       inSub.expectRequest(1)
-      inSub.sendNext(ParserOutput.RequestStart(
-        HttpMethods.GET,
-        Uri("http://example.com/"),
-        HttpProtocols.`HTTP/1.1`,
-        List(),
-        StrictEntityCreator(HttpEntity.Strict(ContentTypes.`application/octet-stream`, ByteString("body"))),
-        true,
-        false))
+      inSub.sendNext(strictRequest)
 
       upstreamProbe.expectNext()
+
+    }
+
+    "not complete running entity stream when upstream cancels" in {
+      implicit val materializer = ActorMaterializer()
+
+      val inProbe = TestPublisher.manualProbe[ParserOutput.RequestOutput]()
+      val upstreamProbe = TestSubscriber.manualProbe[HttpRequest]()
+
+      val stage = Flow.fromGraph(new PrepareRequests(ServerSettings(system)))
+
+      Source.fromPublisher(inProbe)
+        .via(stage)
+        .to(Sink.fromSubscriber(upstreamProbe))
+        .withAttributes(Attributes.inputBuffer(1, 1))
+        .run()
+
+      val upstreamSub = upstreamProbe.expectSubscription()
+      val inSub = inProbe.expectSubscription()
+
+      // let request with streamed entity through
+      upstreamSub.request(1)
+      inSub.expectRequest(1)
+      inSub.sendNext(chunkedStart)
+
+      val request = upstreamProbe.expectNext()
+
+      // and subscribe to it's streamed entity
+      val entityProbe = TestSubscriber.manualProbe[ByteString]()
+      request.entity.dataBytes.to(Sink.fromSubscriber(entityProbe))
+        .withAttributes(Attributes.inputBuffer(1, 1))
+        .run()
+
+      val entitySub = entityProbe.expectSubscription()
+
+      // user logic cancels flow
+      upstreamSub.cancel()
+
+      // but incoming chunks should still end up in entity
+      entitySub.request(1)
+      inSub.expectRequest(1)
+      inSub.sendNext(chunkPart)
+      entityProbe.expectNext()
+
+      entitySub.request(1)
+      inSub.expectRequest(1)
+      inSub.sendNext(chunkRequestComplete)
+
+      // and then when entity is complete, the stage should complete
+      entityProbe.expectComplete()
+
+    }
+
+    "complete stage if chunked stream is completed without reaching end of chunks" in {
+      // a bit unsure about this, but to document the assumption
+      implicit val materializer = ActorMaterializer()
+
+      val inProbe = TestPublisher.manualProbe[ParserOutput.RequestOutput]()
+      val upstreamProbe = TestSubscriber.manualProbe[HttpRequest]()
+
+      val stage = Flow.fromGraph(new PrepareRequests(ServerSettings(system)))
+
+      Source.fromPublisher(inProbe)
+        .via(stage)
+        .to(Sink.fromSubscriber(upstreamProbe))
+        .withAttributes(Attributes.inputBuffer(1, 1))
+        .run()
+
+      val upstreamSub = upstreamProbe.expectSubscription()
+      val inSub = inProbe.expectSubscription()
+
+      // let request with streamed entity through
+      upstreamSub.request(1)
+      inSub.expectRequest(1)
+      inSub.sendNext(chunkedStart)
+
+      val request = upstreamProbe.expectNext()
+
+      // and subscribe to it's streamed entity
+      val entityProbe = TestSubscriber.manualProbe[ByteString]()
+      request.entity.dataBytes.to(Sink.fromSubscriber(entityProbe))
+        .withAttributes(Attributes.inputBuffer(1, 1))
+        .run()
+
+      val entitySub = entityProbe.expectSubscription()
+
+      // incoming stream is completed, but we never got the chunk end
+      inSub.sendComplete()
+
+      // assumption: should cause stage to complete
+      entityProbe.expectComplete()
+      upstreamProbe.expectComplete()
 
     }
   }

